File: TI1/TI1.3 RESTFUL_JAVA_SINBDD_GR09/02.CLICON/CONUNI_RESTFUL_JAVA_CLICON_G09/src/main/java/ec/edu/monster/vista/VistaLogin.java
package ec.edu.monster.vista;

import java.io.Console;
import java.util.Scanner;

/**
 * Vista para el proceso de autenticación - Patrón MVC
 * Maneja la presentación del login y entrada de credenciales
 * @author ACER NITRO V15
 */
public class VistaLogin {
    
    // Colores para la consola
    public static final String CELESTE = "\033[38;5;51m";
    public static final String VERDE = "\033[92m";
    public static final String ROJO = "\033[91m";
    public static final String AMARILLO = "\033[93m";
    public static final String RESET = "\033[0m";
    
    private Scanner scanner;
    private boolean soportaColores;
    
    /**
     * Detecta si el terminal soporta colores
     */
    private boolean detectarSoporteColores() {
        try {
            String os = System.getProperty("os.name").toLowerCase();
            
            if (os.contains("windows")) {
                // En Windows 10/11, verificar si VirtualTerminalLevel está habilitado
                // O simplemente intentar siempre, ya que Windows 10+ lo soporta
                try {
                    // Verificar si estamos en Windows Terminal, PowerShell o CMD con ANSI habilitado
                    String term = System.getenv("TERM");
                    String wtSession = System.getenv("WT_SESSION");
                    
                    // Si hay WT_SESSION, estamos en Windows Terminal
                    if (wtSession != null) {
                        return true;
                    }
                    
                    // Si hay TERM configurado, usarlo
                    if (term != null && (term.contains("xterm") || term.contains("256color"))) {
                        return true;
                    }
                    
                    // Windows 10 build 1511+ soporta ANSI si está habilitado
                    // Intentar siempre en Windows, el sistema lo manejará
                    return true;
                } catch (Exception e) {
                    // Si falla, intentar de todas formas
                    return true;
                }
            }
            
            // Para Linux/Mac, verificar TERM
            String term = System.getenv("TERM");
            return term != null && !term.equals("dumb");
        } catch (Exception e) {
            // En caso de error, intentar usar colores de todas formas
            return true;
        }
    }
    
    /**
     * Aplica color solo si el terminal lo soporta
     */
    private String aplicarColor(String color, String texto) {
        return soportaColores ? color + texto + RESET : texto;
    }
    
    public VistaLogin() {
        // Detectar soporte de colores
        this.soportaColores = detectarSoporteColores();
        
        // Crear Scanner con codificación UTF-8
        try {
            this.scanner = new Scanner(System.in, "UTF-8");
        } catch (Exception e) {
            this.scanner = new Scanner(System.in);
        }
    }
    
    /**
     * Muestra el banner de login
     */
    public void mostrarBannerLogin() {
        mostrarLogo();
        System.out.println("=".repeat(60));
        System.out.println(aplicarColor(CELESTE, "MONSTERS INC. CONVERTER RESTFUL - AUTENTICACIÓN"));
        System.out.println(aplicarColor(CELESTE, "Sistema de Seguridad Empresarial"));
        System.out.println("=".repeat(60));
        System.out.println();
        System.out.println(aplicarColor(AMARILLO, "Acceso restringido - Ingrese sus credenciales"));
        System.out.println();
    }
    
    /**
     * Muestra el logo ASCII de MONSTERS INC
     */
    private void mostrarLogo() {
        System.out.println(
            "\n" +
            aplicarColor(CELESTE, "███╗   ███╗ ██████╗ ███╗   ██╗███████╗████████╗███████╗██████╗     ██████╗ ██████╗ ██████╗") + "\n" +
            aplicarColor(CELESTE, "████╗ ████║██╔═══██╗████╗  ██║██╔════╝╚══██╔══╝██╔════╝██╔══██╗   ██╔════╝ ██╔══██╗╚═══██╗") + "\n" +
            aplicarColor(CELESTE, "██╔████╔██║██║   ██║██╔██╗ ██║███████╗   ██║   █████╗  ██████╔╝   ██║  ███╗██████╔╝  ███╔╝") + "\n" +
            aplicarColor(CELESTE, "██║╚██╔╝██║██║   ██║██║╚██╗██║╚════██║   ██║   ██╔══╝  ██╔══██╗   ██║   ██║██╔══██╗ ███╔╝") + "\n" +
            aplicarColor(CELESTE, "██║ ╚═╝ ██║╚██████╔╝██║ ╚████║███████║   ██║   ███████╗██║  ██║   ╚██████╔╝██║  ██║██████╗") + "\n" +
            aplicarColor(CELESTE, "╚═╝     ╚═╝ ╚═════╝ ╚═╝  ╚═══╝╚══════╝   ╚═╝   ╚══════╝╚═╝  ╚═╝    ╚═════╝ ╚═╝  ╚═╝╚═════╝") + "\n"
        );
    }
    
    /**
     * Lee el nombre de usuario
     */
    public String leerUsuario() {
        System.out.print(aplicarColor(CELESTE, "Usuario: "));
        return scanner.nextLine().trim();
    }
    
    /**
     * Lee la contraseña oculta con asteriscos
     */
    public String leerContrasenaOculta() {
        System.out.print(aplicarColor(CELESTE, "Contraseña: "));
        
        try {
            // Intentar usar Console para ocultar la entrada
            Console console = System.console();
            if (console != null) {
                char[] passwordArray = console.readPassword();
                return new String(passwordArray);
            } else {
                // Fallback: usar entrada normal pero sin mostrar el mensaje feo
                return scanner.nextLine().trim();
            }
        } catch (Exception e) {
            // Si falla, usar entrada normal
            return scanner.nextLine().trim();
        }
    }
    
    /**
     * Muestra mensaje de éxito
     */
    public void mostrarMensajeExito() {
        System.out.println();
        System.out.println(aplicarColor(VERDE, "¡Autenticación exitosa!"));
        System.out.println(aplicarColor(VERDE, "Bienvenido al sistema Monsters Inc. Converter RESTful"));
        System.out.println("=".repeat(60));
        System.out.println();
    }
    
    /**
     * Muestra mensaje de error
     */
    public void mostrarMensajeError(int intentoActual, int maxIntentos) {
        System.out.println();
        System.out.println(aplicarColor(ROJO, "Credenciales incorrectas"));
        
        if (intentoActual < maxIntentos) {
            System.out.println(aplicarColor(AMARILLO, "Intente nuevamente..."));
            System.out.println();
        }
    }
    
    /**
     * Muestra mensaje de bloqueo
     */
    public void mostrarMensajeBloqueo() {
        System.out.println();
        System.out.println(aplicarColor(ROJO, "ACCESO DENEGADO"));
        System.out.println(aplicarColor(ROJO, "Demasiados intentos fallidos"));
        System.out.println(aplicarColor(ROJO, "Contacte al administrador del sistema"));
        System.out.println("=".repeat(60));
        System.out.println();
    }
    
    /**
     * Muestra información sobre el intento actual
     */
    public void mostrarIntento(int intento, int maxIntentos) {
        System.out.println(aplicarColor(AMARILLO, "INTENTO " + intento + " de " + maxIntentos));
        System.out.println();
    }
    
    /**
     * Cierra los recursos
     */
    public void cerrar() {
        if (scanner != null) {
            scanner.close();
        }
    }
}

File: Proyectos/TI.1 PAO202650_SINBDD_NRC30742/TI.1.4. PAO202650_RESTFUL_DOTNET_SINBDD_NRC30742/04. CLICON/CONUNI_RESTFUL_DOTNET_CLICON_G05/CONUNI_RESTFUL_DOTNET_CLICON_G09/vista/VistaLogin.cs
using System;

namespace CONUNI_RESTFUL_DOTNET_CLICON_G09.vista
{
    /// <summary>
    /// Vista para el proceso de autenticación - Patrón MVC
    /// Maneja la presentación del login y entrada de credenciales
    /// </summary>
    public class VistaLogin
    {
        // Colores para la consola (códigos ANSI)
        public const string CELESTE = "\u001b[38;5;51m";
        public const string VERDE = "\u001b[92m";
        public const string ROJO = "\u001b[91m";
        public const string AMARILLO = "\u001b[93m";
        public const string RESET = "\u001b[0m";
        
        private bool soportaColores;

        /// <summary>
        /// Detecta si el terminal soporta colores
        /// </summary>
        private bool DetectarSoporteColores()
        {
            try
            {
                // En Windows 10+, PowerShell y CMD modernos soportan colores ANSI
                string term = Environment.GetEnvironmentVariable("TERM");
                string os = Environment.OSVersion.ToString();
                
                // Habilitar colores ANSI en Windows
                if (Environment.OSVersion.Platform == PlatformID.Win32NT)
                {
                    // Intentar habilitar colores ANSI en Windows
                    try
                    {
                        var handle = System.Diagnostics.Process.GetCurrentProcess().MainWindowHandle;
                        // Windows 10+ soporta ANSI por defecto
                        return true;
                    }
                    catch
                    {
                        return false;
                    }
                }
                
                return term != null && !term.Equals("dumb");
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        /// Aplica color solo si el terminal lo soporta
        /// </summary>
        private string AplicarColor(string color, string texto)
        {
            return soportaColores ? color + texto + RESET : texto;
        }

        public VistaLogin()
        {
            // Detectar soporte de colores
            this.soportaColores = DetectarSoporteColores();
            
            // Habilitar colores ANSI en Windows si es posible
            if (Environment.OSVersion.Platform == PlatformID.Win32NT)
            {
                try
                {
                    // Intentar habilitar modo ANSI
                    Console.OutputEncoding = System.Text.Encoding.UTF8;
                }
                catch { }
            }
        }

        /// <summary>
        /// Muestra el banner de login
        /// </summary>
        public void MostrarBannerLogin()
        {
            MostrarLogo();
            Console.WriteLine(new string('=', 60));
            Console.WriteLine(AplicarColor(CELESTE, "MONSTERS INC. CONVERTER RESTFUL - AUTENTICACIÓN"));
            Console.WriteLine(AplicarColor(CELESTE, "Sistema de Seguridad Empresarial"));
            Console.WriteLine(new string('=', 60));
            Console.WriteLine();
            Console.WriteLine(AplicarColor(AMARILLO, "Acceso restringido - Ingrese sus credenciales"));
            Console.WriteLine();
        }

        /// <summary>
        /// Muestra el logo ASCII de MONSTERS INC
        /// </summary>
        private void MostrarLogo()
        {
            Console.WriteLine();
            Console.WriteLine(AplicarColor(CELESTE, "███╗   ███╗ ██████╗ ███╗   ██╗███████╗████████╗███████╗██████╗     ██████╗ ██████╗ ██████╗"));
            Console.WriteLine(AplicarColor(CELESTE, "████╗ ████║██╔═══██╗████╗  ██║██╔════╝╚══██╔══╝██╔════╝██╔══██╗   ██╔════╝ ██╔══██╗╚═══██╗"));
            Console.WriteLine(AplicarColor(CELESTE, "██╔████╔██║██║   ██║██╔██╗ ██║███████╗   ██║   █████╗  ██████╔╝   ██║  ███╗██████╔╝  ███╔╝"));
            Console.WriteLine(AplicarColor(CELESTE, "██║╚██╔╝██║██║   ██║██║╚██╗██║╚════██║   ██║   ██╔══╝  ██╔══██╗   ██║   ██║██╔══██╗ ███╔╝"));
            Console.WriteLine(AplicarColor(CELESTE, "██║ ╚═╝ ██║╚██████╔╝██║ ╚████║███████║   ██║   ███████╗██║  ██║   ╚██████╔╝██║  ██║██████╗"));
            Console.WriteLine(AplicarColor(CELESTE, "╚═╝     ╚═╝ ╚═════╝ ╚═╝  ╚═══╝╚══════╝   ╚═╝   ╚══════╝╚═╝  ╚═╝    ╚═════╝ ╚═╝  ╚═╝╚═════╝"));
            Console.WriteLine();
        }

        /// <summary>
        /// Lee el nombre de usuario
        /// </summary>
        public string LeerUsuario()
        {
            Console.Write(AplicarColor(CELESTE, "Usuario: "));
            return Console.ReadLine()?.Trim() ?? "";
        }

        /// <summary>
        /// Lee la contraseña oculta con asteriscos
        /// </summary>
        public string LeerContrasenaOculta()
        {
            Console.Write(AplicarColor(CELESTE, "Contraseña: "));
            
            string contraseña = "";
            ConsoleKeyInfo key;

            do
            {
                key = Console.ReadKey(true);

                if (key.Key != ConsoleKey.Backspace && key.Key != ConsoleKey.Enter)
                {
                    contraseña += key.KeyChar;
                    Console.Write("*");
                }
                else if (key.Key == ConsoleKey.Backspace && contraseña.Length > 0)
                {
                    contraseña = contraseña.Substring(0, contraseña.Length - 1);
                    Console.Write("\b \b");
                }
            }
            while (key.Key != ConsoleKey.Enter);

            Console.WriteLine();
            return contraseña;
        }

        /// <summary>
        /// Muestra mensaje de éxito
        /// </summary>
        public void MostrarMensajeExito()
        {
            Console.WriteLine();
            Console.WriteLine(AplicarColor(VERDE, "¡Autenticación exitosa!"));
            Console.WriteLine(AplicarColor(VERDE, "Bienvenido al sistema Monsters Inc. Converter RESTful"));
            Console.WriteLine(new string('=', 60));
            Console.WriteLine();
        }

        /// <summary>
        /// Muestra mensaje de error
        /// </summary>
        public void MostrarMensajeError(int intentoActual, int maxIntentos)
        {
            Console.WriteLine();
            Console.WriteLine(AplicarColor(ROJO, "Credenciales incorrectas"));
            
            if (intentoActual < maxIntentos)
            {
                Console.WriteLine(AplicarColor(AMARILLO, "Intente nuevamente..."));
                Console.WriteLine();
            }
        }

        /// <summary>
        /// Muestra mensaje de bloqueo
        /// </summary>
        public void MostrarMensajeBloqueo()
        {
            Console.WriteLine();
            Console.WriteLine(AplicarColor(ROJO, "ACCESO DENEGADO"));
            Console.WriteLine(AplicarColor(ROJO, "Demasiados intentos fallidos"));
            Console.WriteLine(AplicarColor(ROJO, "Contacte al administrador del sistema"));
            Console.WriteLine(new string('=', 60));
            Console.WriteLine();
        }

        /// <summary>
        /// Muestra información sobre el intento actual
        /// </summary>
        public void MostrarIntento(int intento, int maxIntentos)
        {
            Console.WriteLine(AplicarColor(AMARILLO, "INTENTO " + intento + " de " + maxIntentos));
            Console.WriteLine();
        }
    }
}

